Preliminary NCI test environment

uchchwhash opened the pull request:
Tools to maintain a more persistent test database
than those created for integration tests.

- Add command `test-env setup` and `test-env teardown`
  that sets up and tears down the test database respectively
- Add command `test-env migrate` that indexes datasets
  from the production database that match the given query
  into the test database
- Add test for `metadata_type` inconsistency in ingestion

omad committed:
Ignore mypy cache dir

uchchwhash committed:
Preliminary NCI test environment

Tools to maintain a more persistent test database
than those created by integration tests.

- Add command `test-env setup` and `test-env teardown`
  that sets up and tears down the test database respectively
- Add command `test-env migrate` that indexes datasets
  from the production database that match the given query
  into the test database
- Add test for `metadata_type` inconsistency in ingestion

uchchwhash committed:
Preliminary NCI test environment

Tools to maintain a more persistent test database
than those created by integration tests.

- Add command `test-env setup` and `test-env teardown`
  that sets up and tears down the test database respectively
- Add command `test-env migrate` that indexes datasets
  from the production database that match the given query
  into the test database
- Add test for `metadata_type` inconsistency in ingestion

uchchwhash committed:
Preliminary NCI test environment

Tools to maintain a more persistent test database
than those created by integration tests.

- Add command `test-env setup` and `test-env teardown`
  that sets up and tears down the test database respectively
- Add command `test-env migrate` that indexes datasets
  from the production database that match the given query
  into the test database
- Add test for `metadata_type` inconsistency in ingestion

uchchwhash committed:
Preliminary NCI test environment

Tools to maintain a more persistent test database
than those created by integration tests.

- Add command `test-env setup` and `test-env teardown`
  that sets up and tears down the test database respectively
- Add command `test-env migrate` that indexes datasets
  from the production database that match the given query
  into the test database
- Add test for `metadata_type` inconsistency in ingestion

jeremyh commented on digitalearthau/test_env.py:
```diff
@@ -0,0 +1,354 @@
+
+    class LocalConfig(ConfigBase):
+        """
+        Backporting the environment feature in `develop` to `release-1.5`.
```
Audacious 🙂 

It's probably useful enough that it might be worth actually putting in release-1.5.

jeremyh commented on digitalearthau/config/datacube.conf:
```diff
@@ -0,0 +1,9 @@
+db_port: 6432
+db_database: datacube
+
+[NCI-test]
```
These commands look useful enough that individual developers would probably run them for their own cubes.  Perhaps we could default the `db_database` to the current username, or something unique, so that they can run it out-of the-box without risk of conflicting? (current_username is typical postgres convention for a user's default database)

jeremyh commented on digitalearthau/test_env.py:
```diff
@@ -0,0 +1,354 @@
+
+    # seems like you have to connect to a database
+    # and that the maintenance database is usually called postgres
+    return run_shell(["psql", "-h", hostname, "-p", port, maintenance_db],
```
If this script keeps growing it would be worth using one of the Python Postgres apis rather than calling to a shell command. Such as the Index object's own connection, or plain psycopg2. 

The commands calling this are manually formatting SQL strings, but it would be better to use something that escapes arguments for us to avoid issues with special characters. This is pretty insignificant at the moment (all arguments come from config files, and we're not using special characters in our database names), but everything has a habit of growing beyond its original usage 🙂

jeremyh commented on digitalearthau/test_env.py:
```diff
@@ -0,0 +1,354 @@
+        yield from datasets
+
+
+def normalize_uri(uri):
```
The property `dataset.local_path` could simplify these two methods — the indexing function we use here doesn't currently support uris anyway, and you only need the latest path, I think? We've got too many uri-to-path implementations already 😛

jeremyh commented on integration_tests/test_config.py:
```diff
@@ -66,3 +66,14 @@ def test_dea_config(dea_index: Index):
     }
+
+
+def test_metadata_type(dea_index):
```
👍

jeremyh commented on setup.py:
```diff
@@ -59,6 +59,7 @@
             'dea-submit-sync = digitalearthau.sync.submit_job:main',
             'dea-sync = digitalearthau.sync:cli',
             'dea-system = digitalearthau.system:cli',
+            'test-env = digitalearthau.test_env:cli'
```
This command gets added to everyone's global path, so we don't want command names to be too generic (and possibly conflict with others). 

There's also nothing specifically test-related about it: it's creating a real datacube from another. Perhaps call it `dea-env`? (renaming this module and the above application_name too)

```
   dea-env setup
   dea-env migrate
   dea-env teardown
```


To take the idea further, and don't feel pressured to do this in this pull request, `setup` could be moved into `dea-system`, which currently has an `init` command, but could explicitly create them too.

Something like:
```
 dea-system create
 dea-import  --from-env prod <search_args>
 dea-system drop
```

jeremyh commented on digitalearthau/test_env.py:
```diff
@@ -0,0 +1,354 @@
+    #   target: /g/data/u46/users/ia1511/data/datacube/
+    # or something similar
+
+    # there are too many DuplicateRecordError warnings
```
Those warnings bug me too; it's normal behaviour to add a dataset with duplicate ancestor datasets of another. We should add an issue.

jeremyh reviewed (commented):
Looks good Imam. I'm slightly hesitant about the amount of duplicated functionality from elsewhere, as it's more to maintain in the future (ideally we should improve the upstream methods 🙂 ), but it's not a blocker in the short term.

jeremyh committed:
Make sync test data usable by other tests

jeremyh committed:
Model object shouldn't use Test* naming conventions

PyTest is trying to scan it for tests, outputting a warning.

jeremyh committed:
Represent collections as named tuples for consistency

omad committed:
Don't pin pylint to 1.6.4, it's broken

jeremyh committed:
Expand module docs for serialise, events (#39)

As suggested in #32

jeremyh committed:
Allow test logging setup to be reused

jeremyh committed:
Expand help and comments for move, sync clis

jeremyh committed:
Remove unnecessary index proxy class

This was added to make unit testing easier without a real index.
But we now have a real index to test against in the integration_tests,
and the InMemoryIndex is no longer used much. Better to simplify the
code.

jeremyh committed:
Simpify sync tests, moving index logic into framework

jeremyh committed:
Rename test dataset to clarify intent

jeremyh committed:
Add a test of a basic dataset file move

jeremyh committed:
Add accidentally-ignored test package files

They are listed in the checksum of the package, but gitignore
is ignoring all *.log files.

jeremyh committed:
Add failing test for moving dataset that already exists

omad committed:
Stop installing stats into the `agdc-py3` module

We want to be able to release different code bases independently,
and it's the wrong place to roll it in.

It's now available as an automatic deployment that goes into `agdc_stats`

jeremyh committed:
Logged jsonl should be in compact format

jeremyh committed:
Do cleanup per-product; display progressbar in tty

jeremyh committed:
Automatically record cleanup output in work folder

jeremyh committed:
Support structlog output to file

jeremyh committed:
Record each product cleanup to relevant work directory

jeremyh committed:
Record user arguments in cleanup output

jeremyh committed:
Keep echo statements within cli function

jeremyh committed:
Expand click help for all dea collection tools

jeremyh committed:
Add bare-bones collection management overview

jeremyh committed:
Slight CM page rewording for clarity

jeremyh committed:
Fix sync cli argument reference

omad committed:
Add a Stacker executor that follows the basic model of the FC executor

jeremyh committed:
Shutdown multiproc pools cleanly

By default, the "with" context will terminate, rather than close,
the pool.

omad committed:
Make several stacker fixes

omad committed:
Make dea-stacker work with old style config files

Ones that specify [source|output]_type instead of _product

omad committed:
Import stacker cli docs

omad committed:
Don't install pathlib/pathlib2 in Python 3.6 env

pathlib in particular includes an outdated API, which was giving me
major grief.

pathlib2 is probably harmless, but completely unnecessary.

omad committed:
Fix creation of work directory to use passed in job type

omad committed:
Correctly pass query params in stacker generate

omad committed:
Simplify tag handling

The TaskApp handling stuff in DEA was creating a task_datetime and using
that for the work directory, but other bits of code were expecting a tag
to be passed in on every invocation.

For now, lets go with the TaskApp task_datetime, and use that to generate
a task_tag.

omad committed:
Fix dea-stacker typo

omad committed:
Tiny stacker cleanup

- Remove the --load-tasks option from run(), it's discovered by loading the task description from --task-description
- Remove unnecessary *args, **kwargs. I think click is smart enough to not pass arguments a function can't accept

omad committed:
Make qsub param to decorated functions optional

omad committed:
Must handle having a qsub param on dea-stacker:run()

omad committed:
Add DEA + ODC version numbers for stacker to use

omad committed:
Fix str/path mismatch with odc-stacker code

omad committed:
Estimate required runtime for PBS jobs

omad committed:
Finalise goalkeeper instructions

- Indent headings to avoid polluting the global TOC
- Fix links to release process docs
- Include CI build status images inline in the docs

Closes #31

Commit:
Remove old item/hltc dataset prepare script

Dataset prepare scripts are only appropriate for external datasets,
datasets that we create ourselves should come with metadata embedded.

omad committed:
Fix dea-orchestration travis shield in goalkeeper docs

It was pointing to the wrong default branch

omad committed:
Add note about required modules when creating products

jeremyh committed:
Start a common integration test framework

Allows the database & config initialisation code for
integration testingto be reused by other repositories.

omad committed:
Prepare for FC shuffling, move _staging to active, and archive current

omad committed:
Add support for moving *.nc files with `dea-move`

omad committed:
Start work on splitting collections into a configuration file/database table

- Make the type annotations more accurate (Makes pycharm happier)
- Only Enums with UPPERCASE names are able to be de/serialised by our serialise module

Put index back into Collection

jeremyh committed:
Add integration pylint disables to testing class.

jeremyh committed:
Consistent testing declarations

omad committed:
Cleaning and tidying

omad committed:
Rename for clarity of purpose

omad committed:
Fix copy-o bug in test fixture

omad committed:
Add/fix multiple paths for old and new FC collections

omad committed:
Let dea-move handle .nc datasets

omad committed:
Add an as yet unused example fc dataset

omad committed:
PyYaml is being pernickety

omad committed:
Install pycodestyle instead of pep8

It's the same thing, but with a new name. The Old name will be discontinued

omad committed:
Final clean ups of move improvements

omad committed:
Remove temporary files on NetCDF move failure

uchchwhash committed:
Capture CLI `--qsub` argument

The function `capture_qsub` is supposed to return the supplied value of the `--qsub` parameter in addition to storing it in the state object. It did in its original implementation in the `agdc_statistics` but somehow after the port it is not doing that anymore.

- Return the captured `value` for the `--qsub` parameter from `capture_qsub`

jeremyh committed:
Publish DEA modules to the public directory

jeremyh committed:
Fix reference to dea-duplicates command

CEKrause committed:
Added info about joining DEA Slack

uchchwhash committed:
increase redis memory (#58)

uchchwhash committed:
Add option for specifying number of workers per node (#56)

Some tasks can utilize multiple cores effectively such as Dale Roberts'
geomedian and SMAD routines. Add an option where only a specified
number of workers are spawned per node.

- Add option `--workers-per-node` to the CLI
- Add log messages to facilitate debugging

jeremyh committed:
Remove outdated stats ansible playbook

The current one is in the stats repo.

uchchwhash committed:
another config file is not a good idea

uchchwhash committed:
another config file is not a good idea

uchchwhash committed:
another config file is not a good idea

uchchwhash committed:
another config file is not a good idea

uchchwhash committed:
Merge branch 'NCI-test' of github.com:GeoscienceAustralia/digitalearthau into NCI-test

uchchwhash committed:
add fc products

uchchwhash committed:
add fc products

uchchwhash committed:
fix yaml whitespace

uchchwhash committed:
fix yaml whitespace

uchchwhash committed:
fix test

uchchwhash committed:
fix test

uchchwhash committed:
minor fix

uchchwhash committed:
minor fix

jeremyh committed:
Add new cubedash deps

uchchwhash committed:
Add auto_clean option for self-submission (#60)

- Fix `QSubLauncher` API to enable specifying CLI options
  to remove when self-submitting

Kirill888 committed:
Add `CPL_ZIP_ENCODING` environment variable for passthru (#62)

Needed to work around some gdal issue

uchchwhash committed:
Increase memory margin per node (#61)

The memory left to the rest of the system per node (currently 512MB)
is perhaps not enough to make sure the job gets scheduled faster
according to @daleroberts.

- Change the memory request to be 1GB less per node than available
- Fix tests to reflect the change

jeremyh committed:
Add click-datetime to environment

Requested by Simon O

uchchwhash committed:
test-env

uchchwhash commented:
Notwithstanding the fact that I have not worked through all of @jeremyh's suggestions here, I don't think it would be desirable to merge this branch as it is here. Instead, I am starting anew at the `NCI-test-env` branch that targets core `develop` branch soon to be released as release 1.5.6.
